File: 백준/Gold/14938. 서강그라운드/서강그라운드.cc
#include <bits/stdc++.h>
using namespace std;

const int INF = 0x3f3f3f3f;
// 모든 쌍 최단 경로 문제(ASP) 문제
int n, m, r;
int result = 0;
int item_cnt[101];
int d[101][101];

int main(void) {
    cin.tie(0)->sync_with_stdio(false);
    cin >> n >> m >> r;
    for (int i = 1; i <= n; i++) cin >> item_cnt[i];
    for (int i = 1; i <= n; i++) {
        fill(d[i] + 1, d[i] + n + 1, INF);
        d[i][i] = 0;
    }
    while (r--) {
        int u, v, w;
        cin >> u >> v >> w;
        if (d[u][v] > w) d[u][v] = d[v][u] = w;
    }

    for (int k = 1; k <= n; k++) {
        for (int i = 1; i <= n; i++) {
            for (int j = 1; j <= n; j++) {
                if (d[i][j] > d[i][k] + d[k][j]) {
                    d[i][j] = d[i][k] + d[k][j];
                }
            }
        }
    }

    for (int i = 1; i <= n; i++) {
        int cnt = 0;
        for (int j = 1; j <= n; j++) {
            if (d[i][j] <= m) cnt += item_cnt[j];
        }
        if (result < cnt) result = cnt;
    }

    cout << result;
}

File: 백준/Gold/14938. 서강그라운드/서강그라운드.java
import java.io.*;
import java.util.*;

public class Main {

    // 최대 O(n^3)
    // a -> b 까지 직접적인 길이 없어도 우회해서 갈 수 있음
    // 길이 여러 개 있을 수도 있음 -> 최단 경로를 알아야 함.
    // ASP를 구해서 수색 범위 인지 파악 -> 아이템 개수 파악

    static int n, m, r;
    static int itemCounts[] = new int[103];
    static int dist[][] = new int[103][103];
    static final int INF = 0x3f3f3f3f;

    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        n = Integer.parseInt(st.nextToken());
        m = Integer.parseInt(st.nextToken());
        r = Integer.parseInt(st.nextToken());

        for (int i = 1; i <= n; i++) {
            Arrays.fill(dist[i], 1, n + 1, INF);
            dist[i][i] = 0;
        }

        st = new StringTokenizer(br.readLine());
        for (int i = 1; i <= n; i++) {
            itemCounts[i] = Integer.parseInt(st.nextToken());
        }

        while (r-- > 0) {
            st = new StringTokenizer(br.readLine());
            int u = Integer.parseInt(st.nextToken());
            int v = Integer.parseInt(st.nextToken());
            int w = Integer.parseInt(st.nextToken());
            dist[u][v] = Math.min(dist[u][v], w);
            dist[v][u] = Math.min(dist[v][u], w);
        }

        for (int k = 1; k <= n; k++) {
            for (int i = 1; i <= n; i++) {
                for (int j = 1; j <= n; j++) {
                    dist[i][j] = Math.min(dist[i][j], dist[i][k] + dist[k][j]);
                }
            }
        }

        int maxCnt = 0;
        for (int i = 1; i <= n; i++) {
            int cnt = 0;
            for (int j = 1; j <= n; j++) {
                if (dist[i][j] <= m) {
                    cnt += itemCounts[j];
                }
            }
            maxCnt = Math.max(maxCnt, cnt);
        }
        System.out.print(maxCnt);
    }
}